### ISSUE TITLE
Kafka Consumer GSSAPI Error on CentOS VM #1630](https://github.com/edenhill/librdkafka/issues/1630

### ISSUE THREAD

**Issue by mmgutsche** [cyrusimap/cyrus-sasl#501]
I described a problem here, which might be related to cyrus-sasl. 
I would love to get some information on how to troubleshoot this issue, since it is difficult to replicate and just happens to appear on my virtual machine with CentOS.

Details can be found here: [Kafka Consumer GSSAPI Error on CentOS VM #1630](https://github.com/edenhill/librdkafka/issues/1630).

**Comment by Jakuje** [comment]
CentOS 7.4 has already the patch for "correct" SPNEGO, which makes it fail against the old implementations of SPNEGO in cyrus-sasl:

https://git.centos.org/blob/rpms!cyrus-sasl.git/c7/SPECS!cyrus-sasl.spec#L443

This is probably the main difference from Ubuntu.

There is similar [discussion bug](https://bugzilla.redhat.com/show_bug.cgi?id=1481467) for Fedora, which got also this patch.

A quick fix is to configure your tools to not offer GSS-SPNEGO or fix the SPNEGO the same way also in the other tool you are talking to.

**Comment by mmgutsche** [comment]
Thanks for the quick reply. I'm by far means no expert in Kerberos, SASL or all these authentication mechanism. So which tool is in charge of offering GSS-SPNEGO? Kerberos? The Kafka Server? The Kafka Client?

**Comment by Jakuje** [comment]
Me neither. But for ldap, which was broken similar way, the change was in configuration file /etc/sasl2/slapd.conf to add "mech_list: gssapi". There should be some similar configuration for either client or server.

**Comment by edenhill** [comment]
@Jakuje I guess that means calling sasl_client_start() with mechlist="GSSAPI"? We already do this in librdkafka (the Kafka client) so I'm not sure that's the issue.

Relevant code (`rk->rk_conf.sasl.mechanisms` is set to just "GSSAPI"):
https://github.com/edenhill/librdkafka/blob/master/src/rdkafka_sasl_cyrus.c#L485

**Comment by Jakuje** [comment]
The issue is probably with the SPNEGO (you might try to recompile it without that patch to confirm it is really the source of the problem or update the server to have this patch also), as we can read through the debug logs, it is listed before the GSSAPI in all the cases from [your code](https://github.com/edenhill/librdkafka/blob/6abae42b0a0313c07278a88c22fe0ea8ebbba81f/src/rdkafka_sasl_cyrus.c#L476):
```
%7|1515506815.492|SASL|rdkafka#consumer-1| [thrd:sasl_plaintext://remote_server:9092/bootstrap]: sasl_plaintext://remote_server:9092/bootstrap: My supported SASL mechanisms: GSS-SPNEGO GSSAPI EXTERNAL LOGIN PLAIN ANONYMOUS
```
I would double-check that the configuration is correctly in place and really effective. Also checking what is the `mech` returned from `sasl_client_start()` could bring some light into the problem.

I am no expert for SASL either so you might get some better tips.

**Comment by mmgutsche** [comment]
It seems that GSSAPI is selected, but we still get the error. Any ideas, or who else could have some ideas?

**Comment by Jakuje** [comment]
I would give it a try with recompiled package without this patch, if something will change. Or just with some older package from older centos, which did not contain this patch. If it will behave the same, problem is elsewhere. If not, we can dig into these changes.

**Comment by mmgutsche** [comment]
So I compiled the following commit: 59bee00
Latest changelog of that commit:
```
2016-10-18  Ken Murchison <murch@andrew.cmu.edu>
	* Fixed potential DoS attack on saslauthd/doors (from Oracle)

2016-06-30  Ken Murchison <murch@andrew.cmu.edu>
	* plugins/ntlm.c, otp.c: support OpenSSL 1.1

2016-06-14  Ken Murchison <murch@andrew.cmu.edu>
	* plugins/digestmd5.c: Fix memory leak in client step 2

2016-03-24  Ken Murchison <murch@andrew.cmu.edu>
```

But the error message is the same. So it seems not related to cyrus-sasl or at least not to that specific patch. Looking at it now from the krb5 side.

**Comment by mmgutsche** [comment]
So after clearing some obstacles with correct krb5 caches we are back at libsasl

> Stlll seems to be an issue with libsasl2, it never appears to return from sasl_client_start() in the CentOS case.
> After this log (which is from libsasl2) we don't see any more output for broker2:
> %5|1516023480.667|LIBSASL|rdkafka#consumer-1| [thrd:sasl_plaintext://host2:9092/bootstrap]: sasl_plaintext://host2:9092/bootstrap: GSSAPI client step 1

Any idea how to find out why it is stuck there?

**Comment by Jakuje** [comment]
It is still hard to say where did it stuck and why. You can find out by attaching `gdb` to the stuck process and investigating the backtrace, if you have a debuginfo installed. It should point you to some fine in the source code.
My guess is some lock, which would mean that there is more processes/threads in some unsafe way, or there is logical error.

Certainly, there must be some difference between your Ubuntu and CentOS machine. Unfortunately, both of the versions are bloated with many patches which makes it very hard to figure out difference on the first sight:

https://anonscm.debian.org/cgit/pkg-cyrus-sasl2/cyrus-sasl2.git/tree/debian/patches?h=debian/2.1.26.dfsg1-14&id=dbc791929151c058e781a445391b221f11178ef8

I would probably start from building the latest pre-release version from cyrus-sasl on both systems and check if it will work at least on Ubuntu. If it does, it is problem, that should get resolved before releasing the new 2.1.27, which is already due. If it will, work, there is "only" a bug in CentOS that will probably need to be addressed, but it will require some more debugging.

**Comment by mmgutsche** [comment]
So I got the cause, see [other issue](https://github.com/edenhill/librdkafka/issues/1630) for details.
